libjuju: headless failure to add machines

```
2017-04-19 21:59:30,157 [DEBUG] conjure-up/_unspecified_spell - app.py:216 - LXD version: 2.12, Juju version: 2.2-beta3-xenial-amd64
2017-04-19 21:59:30,168 [DEBUG] conjure-up/_unspecified_spell - app.py:274 - found spell {'key': 'kubernetes-core', 'name': 'Kubernetes Core', 'description': 'Kubernetes is an open-source platform for deploying, scaling, and operations of application containers across a cluster of hosts. This is a minimal Kubernetes cluster with intended usage for small deployments for testing and development.\n'}
2017-04-19 21:59:30,168 [DEBUG] conjure-up/kubernetes-core - download.py:60 - Path is local filesystem, copying /tmp/tmp.0J38g9higX/spells/kubernetes-core to /home/ubuntu/.cache/conjure-up/kubernetes-core
2017-04-19 21:59:30,173 [INFO] conjure-up/kubernetes-core - events.py:107 - Watching for shutdown
2017-04-19 21:59:30,174 [DEBUG] conjure-up/kubernetes-core - utils.py:404 - Pulling bundle for kubernetes-core from channel: stable
2017-04-19 21:59:32,832 [DEBUG] conjure-up/kubernetes-core - common.py:128 - Found an IPv4 address, assuming LXD is configured.
2017-04-19 21:59:32,832 [INFO] conjure-up/kubernetes-core - common.py:183 - Bootstrapping Juju controller.
2017-04-19 21:59:32,833 [DEBUG] conjure-up/kubernetes-core - juju.py:180 - bootstrap cmd: ['juju', 'bootstrap', 'localhost', 'conjure-up-controller', '--default-model', 'conjure-up-test', '--config', 'image-stream=daily', '--config', 'enable-os-upgrade=false', '--bootstrap-series', 'xenial']
2017-04-19 21:59:32,835 [DEBUG] conjure-up/kubernetes-core - juju.py:189 - waiting for proc
2017-04-19 22:01:40,213 [DEBUG] conjure-up/kubernetes-core - juju.py:191 - proc done
2017-04-19 22:01:40,213 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting Bootstrapped
2017-04-19 22:01:40,214 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting ModelAvailable
2017-04-19 22:01:40,214 [INFO] conjure-up/kubernetes-core - common.py:207 - Bootstrap complete.
2017-04-19 22:01:40,214 [INFO] conjure-up/kubernetes-core - juju.py:128 - Connecting to model conjure-up-controller:conjure-up-test...
2017-04-19 22:01:40,239 [INFO] conjure-up/kubernetes-core - websocket: connection.py:161 - Driver connected to juju wss://10.0.8.183:17070/model/ebf24553-a359-47fc-86ae-d80cca90661c/api
2017-04-19 22:01:40,301 [WARNING] conjure-up/kubernetes-core - websocket: connection.py:444 - Could not find a set of facades for 2.2-beta3.1. Using the latest facade set instead
2017-04-19 22:01:40,324 [INFO] conjure-up/kubernetes-core - websocket: connection.py:161 - Driver connected to juju wss://10.0.8.183:17070/model/ebf24553-a359-47fc-86ae-d80cca90661c/api
2017-04-19 22:01:40,400 [WARNING] conjure-up/kubernetes-core - websocket: connection.py:444 - Could not find a set of facades for 2.2-beta3.1. Using the latest facade set instead
2017-04-19 22:01:40,443 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting ModelConnected
2017-04-19 22:01:40,443 [INFO] conjure-up/kubernetes-core - juju.py:132 - Connected
2017-04-19 22:01:40,599 [INFO] conjure-up/kubernetes-core - common.py:22 - Waiting for model connection
2017-04-19 22:01:40,599 [INFO] conjure-up/kubernetes-core - common.py:24 - Got model connection
2017-04-19 22:01:40,599 [INFO] conjure-up/kubernetes-core - utils.py:125 - Running pre-deploy.
2017-04-19 22:01:40,599 [DEBUG] conjure-up/kubernetes-core - utils.py:134 - Executing script: /home/ubuntu/.cache/conjure-up/kubernetes-core/steps/00_pre-deploy
2017-04-19 22:01:40,675 [INFO] conjure-up/kubernetes-core - utils.py:172 - Finished pre-deploy: Successful pre-deploy.
2017-04-19 22:01:40,675 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting PreDeployComplete
2017-04-19 22:01:40,675 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting Error
2017-04-19 22:01:40,676 [ERROR] conjure-up/kubernetes-core - events.py:96 - Unhandled exception
Traceback (most recent call last):
  File "/snap/conjure-up/200/lib/python3.6/site-packages/conjureup/events.py", line 94, in handle_exception
    raise exc
  File "/snap/conjure-up/200/lib/python3.6/site-packages/conjureup/controllers/deploy/tui.py", line 24, in do_deploy
    msg_cb=utils.info)
  File "/snap/conjure-up/200/lib/python3.6/site-packages/conjureup/juju.py", line 496, in add_machines
    for m in machines),
  File "/snap/conjure-up/200/lib/python3.6/site-packages/conjureup/juju.py", line 496, in <genexpr>
    for m in machines),
TypeError: string indices must be integers
2017-04-19 22:01:40,676 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting Shutdown
2017-04-19 22:01:40,677 [INFO] conjure-up/kubernetes-core - events.py:113 - Shutting down
2017-04-19 22:01:40,677 [INFO] conjure-up/kubernetes-core - events.py:119 - Disconnecting model
2017-04-19 22:01:40,679 [INFO] conjure-up/kubernetes-core - events.py:121 - Disconnected
```

--- a/conjureup/controllers/deploy/gui.py
+++ b/conjureup/controllers/deploy/gui.py
@@ -209,12 +209,11 @@ class DeployController:
             await events.MAASConnected.wait()
         app_placements = self.get_all_assignments(application)
         juju_machines = app.metadata_controller.bundle.machines
-        machines = []
+        machines = {}
         for virt_machine_id, _ in app_placements:
             if virt_machine_id in self.deployed_juju_machines:
                 continue
             machine_attrs = {
-                'virt_machine_id': virt_machine_id,
                 'series': application.csid.series,
             }
             if cloud_type == 'maas':
@@ -222,7 +221,7 @@ class DeployController:
                     await self.get_maas_constraints(virt_machine_id)
             else:
                 machine_attrs.update(juju_machines[virt_machine_id])
-            machines.append(machine_attrs)
+            machines[virt_machine_id] = machine_attrs
 
         return await juju.add_machines([application], machines,
                                        msg_cb=app.ui.set_footer)

--- a/conjureup/controllers/deploy/tui.py
+++ b/conjureup/controllers/deploy/tui.py
@@ -19,7 +19,7 @@ class DeployController:
                               key=attrgetter('service_name'))
 
         await common.pre_deploy(msg_cb=utils.info)
-        await juju.add_machines([a.service_name for a in applications],
+        await juju.add_machines(applications,
                                 machines,
                                 msg_cb=utils.info)
         tasks = []

--- a/conjureup/events.py
+++ b/conjureup/events.py
@@ -87,7 +87,7 @@ def handle_exception(loop, context):
         return  # already reporting an error
     Error.set()
     exc = context['exception']
-    track_exception(exc.args[0])
+    track_exception(str(exc))
 
     # not sure of a cleaner way to log the exception instance
     try:

--- a/conjureup/juju.py
+++ b/conjureup/juju.py
@@ -478,7 +478,7 @@ async def add_machines(applications, machines, msg_cb):
     Arguments:
 
     app: name of app to which the machines belong
-    machines: list of dictionaries of machine attributes.
+    machines: a mapping of virtual machine numbers to machine attributes.
     The key 'series' is required, and 'constraints' is the only other
     supported key
 
@@ -492,14 +492,16 @@ async def add_machines(applications, machines, msg_cb):
     if machines:
         msg = 'Adding machine{}: {}'.format(
             's' if len(machines) > 1 else '',
-            ', '.join(repr((m['series'], m.get('constraints', '')))
-                      for m in machines),
+            ', '.join(
+                '{}: ({}, {})'.format(v, m['series'], m.get('constraints', ''))
+                for v, m in machines.items()),
         )
         app.log.info(msg)
         msg_cb(msg)
 
         tasks = []
-        for machine in machines:
+        for vmid in sorted(machines.keys()):
+            machine = machines[vmid]
             series = machine['series']
             constraints = constraints_to_dict(machine.get('constraints', ''))
             tasks.append(app.juju.client.add_machine(series=series,
@@ -507,8 +509,8 @@ async def add_machines(applications, machines, msg_cb):
 
         created_machines = await asyncio.gather(*tasks)
         ids = {
-            m['virt_machine_id']: c.id
-            for m, c in zip(machines, created_machines)
+            vmid: m.id
+            for vmid, m in zip(sorted(machines.keys()), created_machines)
         }
 
         msg = "Added machine{}: {}".format(

--- a/requirements_test.txt
+++ b/requirements_test.txt
@@ -6,3 +6,4 @@ nose
 tox
 q
 ipython
+ipdb

